### ISSUE TITLE
dataloss / offline cache showing while online

### ISSUE THREAD

**Issue by iturn** [adriancarriger/angularfire2-offline#44]
Sometimes the database seems to load from offline cache while its online. The data is written to the database and received in firebase and reflecting on other clients. But the current client continues to show the outdated data, a hard refresh in chrome seems to fix this. 
Any idea?

**Comment by iturn** [comment]
I fixed the issue via a workaround, probably the issue isnt in angularfire2-offline

**Comment by iturn** [comment]
Issue still persists , maybe its an issue of this package

**Comment by iturn** [comment]
After much debugging the cause data loss was the following.
On slow computers the offfline write takes time to execute. Sometimes a lot like 10 seconds, if a user runs another similar write the promise gets broken somehow.
A callback when offline write comples would be awseome. 
For example like this, although not sure this is the correct place to do this in the package.

//offline-write.js

export function OfflineWrite(firebasePromise, type, ref, method, args, localUpdateService) {
    localUpdateService.update('write', function (writeCache) {
        if (!writeCache) {
            writeCache = {
                lastId: 0,
                cache: {}
            };
        }
        writeCache.lastId++;
        writeCache.cache[writeCache.lastId] = { type: type, ref: ref, method: method, args: args };
        return writeCache;
    }).then(function (writeCache) {
        var id = writeCache.lastId;
        firebasePromise.then(function () {
            WriteComplete(id, localUpdateService);
            
        }).then(console.log("LocalStorageServer:: offline write done"));
    });
}
        var id = writeCache.lastId;
        firebasePromise.then(function () {
            WriteComplete(id, localUpdateService);
            
        }).then(console.log("LocalStorageServer:: offline write done"));

**Comment by adriancarriger** [comment]
Hi @iturn thanks for posting this issue! The easiest way for me to fix this is if I can reproduce it, so I need a little bit more info:

1. Is the app online the entire time?
1. Is the similar write made to the exact same reference?
1. Are you experiencing the issue with a list or object?
1. What kind of write are you making (push, set, etc)?
1. Can you show your code, or show a minimal example that reproduces the issue?

**Comment by iturn** [comment]
Thanks for your awesome work on this repo and your response! 
Im here for any questions or additional info. 

All of the below is related to the slow write on OfflineWrite while CPU is high, or a slow/old computer is used. I use a single core VM with CPU load app to simulate this, but even on this VM with cpu stress its much faster then on a new system with Intel Celeron N3060 of which my client bought 15 :-(. On these slow computers the offline write after calling sometimes takes up to 15 seconds while on my vm approx 3.
I would also like to mention that all of the above is a complete non issue on decent hardware, since the offline write completes fast.

1. The app is loaded online, then internet is disconnected
2. I don't think the writes are to the exact same reference. But to clarify and example and saying this, they are made to children of the same node.
/1234/1234/1234/a : "12"
/1234/1234/1234/b : "13"
/1234/1234/5678/a : "14"
/1234/1234/5678/b : "15"

3-4. its and update object call
5. See the img for the console logs called and the time delay between the 2 on a fast computer.
the object written in the update object call is very simple 4 string properties on a database object with around 15 mixed properties.

**// function in some-service**
  updatePerson(key: string, data):void {
    this.angularFireOfflineDatabase.object(`/projectPersons/${this.personsKey}/${personKey}`)
       .update(data )
       .then(console.log("ProjectDbService:: receveived person update"))
  }

**// offline-write.js in angularfire2-offline package**
export function OfflineWrite(firebasePromise, type, ref, method, args, localUpdateService) {
    localUpdateService.update('write', function (writeCache) {
        if (!writeCache) {
            writeCache = {
                lastId: 0,
                cache: {}
            };
        }
        writeCache.lastId++;
        writeCache.cache[writeCache.lastId] = { type: type, ref: ref, method: method, args: args };
        return writeCache;
    }).then(function (writeCache) {
        var id = writeCache.lastId;
        firebasePromise.then(function () {
            WriteComplete(id, localUpdateService);
            
        }).then(console.log("LocalStorageServer:: offline write done"));
    });
}

**Comment by adriancarriger** [comment]
Thanks for the info @iturn ..I'll be working on reproducing the issue 👍

**Comment by iturn** [comment]
If any help needed with reprodction, please advice.

**Comment by adriancarriger** [comment]
Hi @iturn ..this is a tricky issue to reproduce so I'll just walk you through what I've done so far:

I've attempted to reproduce the issue [here](https://github.com/adriancarriger/angularfire2-offline/blob/gh-pages/issues/44.1/src/app/app.component.ts#L26-L36) ([live demo](https://adriancarriger.github.io/angularfire2-offline/issues/44.1/dist/)).

It runs two queries on sibling references quickly in hope that the second will overwrite the first in this way:

```ts
console.log('starting write #1');
this.objectExample.update({maxSpeed: 100});
setTimeout(() => {
  console.log('starting write #2');
  this.objectExample.update({engine: null});
}, 3);
```
## Steps Taken

Please try these steps to tell me if you get the same results:

1. Visit [the demo](https://adriancarriger.github.io/angularfire2-offline/issues/44.1/dist/)
1. Press the `Reset example (press while online)` button
1. Open chrome dev tools, go to `Application`, then `Clear storage`, and click `Clear site data`
![screen shot 2017-06-21 at 7 39 36 am](https://user-images.githubusercontent.com/7516020/27384533-cee1bcae-5654-11e7-842c-2035f9af2892.png)

1. In dev tools go to `Performance`, then under `CPU:` set the dropdown to `20x slowdown`
![screen shot 2017-06-21 at 7 39 08 am](https://user-images.githubusercontent.com/7516020/27384516-bde727ae-5654-11e7-90bb-1d267def4b55.png)

1. Refresh page
1. Go to dev tools `Application`, then `IndexedDB`
  - This part is weird for me.. sometimes I have to refresh to page for the `IndexedDB` part of Chrome to show anything, and sometimes I have to restart Chrome itself. If you don't see what is in the following image, please refresh or restart Chrome, and it should show.
1. Drill down to `localforage` then `keyvaluepairs` and expand the value related to the key `read/objectcar`. Your screen should look like this:
![screen shot 2017-06-21 at 7 46 11 am](https://user-images.githubusercontent.com/7516020/27384808-e9d91c9a-5655-11e7-89dc-297c7af70330.png)
1. Disconnect from the internet and go to the dev tools console
1. Press `Simulate two writes (press while offline)`
1. In the dev tools console you should see this:
![screen shot 2017-06-21 at 7 50 18 am](https://user-images.githubusercontent.com/7516020/27384904-5063c532-5656-11e7-9f8b-dfdb99fcda81.png)
1. Go to dev tools `Application`, then `IndexedDB` and expand all the values. You should see this:
![screen shot 2017-06-21 at 7 52 44 am](https://user-images.githubusercontent.com/7516020/27385009-b3198bf8-5656-11e7-98ee-25d1058c8e57.png)
1. Go online
1. Go the the console, and you should see this:
![screen shot 2017-06-21 at 7 51 09 am](https://user-images.githubusercontent.com/7516020/27385036-c8926036-5656-11e7-851c-2202245b759d.png)

## Results

Please let me know if you're seeing the same thing as me. In my case here's what's happening

- The `IndexedDB` value for `"write"` has an object with a key called `cache`. It contains all of the writes that are "replayed" when the user goes back online. They are all saved separately. When you make multiple offline writes, a new write should not overwrite a previous write even if it is still getting saved.

- In the demo I've added logging for when an offline write completes. It should also log the internal unique id which will line up with the `cache` key inside the `IndexedDB` value for `"write"`.

- If you see two completion messages with two different internal ids, then you know that your offline writes are probably working correctly.

- Each message will look something like this `Finished offline write (internal id: 1)`

# Next steps

If you you get a message for both writes then you must be losing data during something that occurs after the offline write is complete. Let me know what happens. Thanks!

**Comment by iturn** [comment]
Indeed you are right, i have exactly the same result as you do following your steps.
![windows_7_x64](https://user-images.githubusercontent.com/12299090/27385939-7d2270a8-5694-11e7-96ed-3fcda0583710.png)

Im not sure where to continue debugging this issue, if the conclusion is that the issue is in my application and not in the package

**Comment by iturn** [comment]
although i now activated the cpu loader app and again see this delay
![windows_7_x64](https://user-images.githubusercontent.com/12299090/27386112-15559292-5695-11e7-9f67-e99a6813cf81.png)

**Comment by adriancarriger** [comment]
I hope we're getting closer to solving this! I have a few questions:

- What do you mean by "CPU loader app"?
- What do you mean by "100%"? Is that the CPU throttling? In Chrome 59 it shows me options like `20x slowdown`
- Did you click the button once and get the duplicate messages?
- Can you list all the actions you took step-by-step (especially what you did differently the second time)?

**Comment by iturn** [comment]
This is an example of a CPU loader app for windows; https://www.jam-software.com/heavyload/ 
I dont have much insights in the chrome cpu options. So i applied both 20X slow and the CPU loader app.
I did click the button once without cpu loader, and 1 time with.
the steps are exactly like you described, just 1 variable, the cpu loader app.

**Comment by adriancarriger** [comment]
So what I understand is that while using the CPU loader app it takes 15 seconds to save. When do you experience the dataloss?

**Comment by iturn** [comment]
When the application is closed within that 15 seconds period. Or when
another call is made and the still pending call gets overwritten/deleted.

Kind Regards,
Met vriendelijke groeten

Stijn Post

+31 6 38 75 46 44
www.iturn.it
stijn@iturn.it


On 22 June 2017 at 14:33:21, Adrian Carriger (notifications@github.com)
wrote:

So what I understand is that while using the CPU loader app it takes 15
seconds to save. When do you experience the dataloss?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/adriancarriger/angularfire2-offline/issues/44#issuecomment-310366548>,
or mute the thread

**Comment by adriancarriger** [comment]
I now understand! You would like to warn the user if they are closing the app before the save is complete.

A possible solution would be to add an additional promise for the offline save that you could use like this:

```ts
const promise = this.afoDatabase.object('car').update({maxSpeed: 100});

promise.offline.then(() => console.log('offline write complete!'));
```

Let me know if that would work. Thanks!

**Comment by iturn** [comment]
Yes you are right that is what  would fit perfectly!

Do you mean to implement this promise in your package, or should I implement in my codebase? If you do it, would it work like this? Call a .then() in the service where I do the this.afoDatabase.object('car').update({maxSpeed: 100}); ?

**Comment by adriancarriger** [comment]
Offline writes now work in version 4.2.1!  🎉 

[Demo](https://adriancarriger.github.io/angularfire2-offline/issues/44.2/dist/) (with [Code](https://github.com/adriancarriger/angularfire2-offline/blob/gh-pages/issues/44.2/src/app/app.component.ts#L29-L31))

## Usage

```ts
const promise = this.afoDatabase.object('car').update({maxSpeed: 100});
promise.offline.then(() => console.log('offline write complete!'));
```

**Comment by iturn** [comment]
Great news! Thanks :)

Kind Regards,
Met vriendelijke groeten

Stijn Post

+31 6 38 75 46 44
www.iturn.it
stijn@iturn.it


On 4 August 2017 at 03:10:54, Adrian Carriger (notifications@github.com)
wrote:

Closed #44
<https://github.com/adriancarriger/angularfire2-offline/issues/44>.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/adriancarriger/angularfire2-offline/issues/44#event-1192404245>,
or mute the thread
